rfcs/plugin-api/RFC.md

```diff
@@ -0,0 +1,279 @@
+	)
+	const sdk = await env.$import("https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js")
+
+	const standardLintRules = await env.$import(
+		"https://cdn.jsdelivr.net/gh/inlang/inlang@1/dist/sdk-js/index.js",
+	)
+
+	return {
+		referenceLanguage: "en",
+		plugins: [
+			// @ivanhofer @jannesblobel the "collection" discussion is resolved with this approach.
+			// just use standardLintRules as plugin 🎉
+			standardLintRules(),
+			pluginJson({
+				pathPattern: "./languages/{language}.json",
+			}),
+			sdk({
+				languageNegotiation: {
+					strategies: [{ type: "localStorage" }, { type: "navigator" }],
+				},
+			}),
+		],
+	}
+}
+```
+
+|                       Pros                       |                           Cons                            |
+| :----------------------------------------------: | :-------------------------------------------------------: |
+| Apps can read, extend and modify previous config | Requires a utility functions that executes `defineConfig` |
+|            No deep nesting of plugins            |                                                           |
+|    familiar interface (vite, rollup), good DX    |                                                           |
+|                no breaking change                |                                                           |
+|       flexible, future proof architecture        |                                                           |
+|              Error handling system               |                                                           |
+
+This proposal is simple and flexible.
+
+1. Have a utility function called `createPlugin` [similar to `createLintRule` in the lint rule system](https://github.com/inlang/inlang/blob/main/source-code/core/src/lint/createLintRule.test.ts).
+
+2. Have a `plugins` property that takes an array of plugins in @inlang/core/config:
+
+```ts
+return {
+	referenceLanguage: "en",
+	plugins: [
+		// pluginJson
+		pluginJson({
+			pathPattern: "./languages/{language}.json",
+		}),
+	],
+}``
+```
+
+3. Have a utility function `setupPlugins` that wraps `defineConfig` to merge the configs of all plugins:
+
+```ts
+const config = await setupPlugins(defineConfig(env))
+```
+
+Example Input:
+
+```
+{
+  referenceLanguage: "en",
+  plugins: [
+    [Function: myPlugin]
+  ]
+}
+
+```
+
+Example Output:
+
+_`withPlugins` executes `defineConfig` (which should be executed on loading the inlang config)_
+
+```
+{
+  referenceLanguage: "en",
+  languages: ["en", "de"],
+  readResources: [Function: readResources],
+  writeResources: [Function: writeResources],
+  plugins: [
+    {
+      id: "samuelstroschein.plugin-json",
+      defineConfig: [Function: defineConfig]
+    }
+  ]
+}
+```
+
+**Why this approach seems good**
+
+1. We can extend the `plugins` property to support more complex use cases in the future:
+
+```ts
+type Plugin = {
+	id: string
+	defineConfig: (env: Env) => Promise<Config>
```

Comment:
we also need to pass the existing config to the `defineConfig` function

Reply:
Yes, would be done with `setupPlugins(defineConfig)`. In other words, we would have a "post-process" step after reading the config. The good news: The post-process step could be used for other things as well.

Reply:
On that note, we should not call the post-process/post-import function `setupPlugins` in anticipation of adding other functionality to the function in the future.